[app] Initialise AngularCLI project

zerefwayne opened the pull request:
This PR initialises an Angular CLI Project and the following changes have been made to the boilerplate code:

- Remove boilerplate code
- Install @angular/pwa support
- Refresh PWA Cache on update
- Comment out failing dummy test
- Add @angular/material

Screenshots:

![image](https://user-images.githubusercontent.com/28949397/121160843-11768280-c86a-11eb-9262-6a3747e0ea56.png)

zerefwayne committed:
[app] Initialise AngularCLI project

- Remove boilerplate code
- Install @angular/pwa support
- Refresh PWA Cache on update
- Comment out failing dummy test
- Add @angular/material
- Add Source Code Headers

MahmoudAGawad commented on diagnostics-app/.browserslistrc:
```diff
@@ -0,0 +1,17 @@
+# This file is used by the build system to adjust CSS and JS output to support the specified browsers below.
```
Do we really need this file as part of the repo?

zerefwayne replied:
It was an autogenerated file, removed it. 👍🏻

MahmoudAGawad commented on diagnostics-app/.editorconfig:
```diff
@@ -0,0 +1,16 @@
+# Editor configuration, see https://editorconfig.org
```
same comment: let's remove editor-specific files from here. Maybe add them in the upper level .gitignore

MahmoudAGawad commented on diagnostics-app/README.md:
```diff
@@ -0,0 +1,27 @@
+# DiagnosticsApp
```
Please describe what this folder represents.

Is this the PWA high level folder? if so, please rename it so that it becomes self-descriptive.

zerefwayne replied:
Updated to `Diagnostics Progressive Web App`

MahmoudAGawad commented on diagnostics-app/karma.conf.js:
```diff
@@ -0,0 +1,44 @@
+// Karma configuration file, see link for more information
```
can we move test-related files into a separate folder. e.g. diagnostics-app/test/?

zerefwayne replied:
[Angular Test Style Guide](https://angular.io/guide/testing#test-file-name-and-location
)

According to this, there can be two ways of organising tests in an Angular project. Looks like keeping test files with component has more pros than a separate test folder.

MahmoudAGawad replied:
Thanks for sharing! Let's keep them in the same directory, then.

Also, let's not forget to add a test/ folder when it comes to integration tests, according to this: https://angular.io/guide/testing#place-your-spec-files-in-a-test-folder

MahmoudAGawad commented on diagnostics-app/src/app/app-routing.module.ts:
```diff
@@ -0,0 +1,16 @@
+ * Copyright 2021 The Chromium Authors. All rights reserved.
+ * Use of this source code is governed by a BSD-style license that can be
+ * found in the LICENSE file.
+ */
```
for consistency with Chromium project. Can you change this to //:
```
// Copyright 2021 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
```

Please do this in the other files as well.

MahmoudAGawad commented on diagnostics-app/src/app/app.component.html:
```diff
@@ -0,0 +1,8 @@
+ found in the LICENSE file.
+-->
+
+<h1>Sample Diagnostics App</h1>
```
Chrome OS Diagnostics App

MahmoudAGawad commented on diagnostics-app/src/app/app-routing.module.ts:
```diff
@@ -0,0 +1,16 @@
+ * Use of this source code is governed by a BSD-style license that can be
+ * found in the LICENSE file.
+ */
+
```
In each of these files, please follow Google's style guide: https://google.github.io/styleguide/jsguide.html#source-file-structure
For example, each file should start with @fileoverview {description}, e.g:
```
/**
  * @fileoverview Description of file, its uses and information
  * about its dependencies.
  * Note: Auto-generated file.
  */
```

zerefwayne replied:
Done these changes in PR #4

MahmoudAGawad commented on diagnostics-app/src/app/app.component.spec.ts:
```diff
@@ -0,0 +1,41 @@
+    expect(app).toBeTruthy();
+  });
+
+  // it(`should have as title 'diagnostics-app'`, () => {
+  //   const fixture = TestBed.createComponent(AppComponent);
+  //   const app = fixture.componentInstance;
+  //   expect(app.title).toEqual('diagnostics-app');
+  // });
```
remove if not used.

MahmoudAGawad commented on diagnostics-app/src/test.ts:
```diff
@@ -0,0 +1,31 @@
+/**
```
same comment: would it be better to isolate test-related logic in a separate folder?

MahmoudAGawad replied:
as discussed in the other thread, we are going to follow Angular Style guide, and leave them together in the same folder.

MahmoudAGawad commented on diagnostics-app/src/webcomponents/hello-world/hello-world.element.ts:
```diff
@@ -0,0 +1,15 @@
+import { LitElement, html } from 'lit';
+import { customElement } from 'lit/decorators.js';
+
+@customElement('hello-world')
+export class HelloWorldElement extends LitElement {
+  render() {
+    return html` <h1>Hello World from LitElement!</h1> `;
+  }
```
What is the purpose of this file? It seems like you are testing LitElement? Let's not add unused code.

zerefwayne replied:
Yes, I included it as a dummy component for future references. Removed it. 👍🏻 I'll make sure to not push unused code from next time.

MahmoudAGawad reviewed (commented):
Hi Aayush,

Thanks for the PR. It looks reasonable but I left comments.

More comments:
1. I see this file is empty: diagnostics-app/src/assets/.gitkeep, What is this for? Can you remove  it, if not going to be used?
2. What are the .png and icons? Are these Angular-logo? We need to use icons with appropriate copyrights. Let me know if you need help with this.

Thanks!

MahmoudAGawad commented:
One more comment.

This PR is already large. Can we focus on one or two goals only? I see there are many goals here:

Remove boilerplate code
Install @angular/pwa support
Refresh PWA Cache on update --> move to a following PR
Comment out failing dummy test
Add @angular/material
Add source code headers
Add WebComponent support and Lit --> move to a following PR
Add Light and Dark theme functionality --> move to a following PR

zerefwayne commented:
@MahmoudAGawad 

> 1. I see this file is empty: diagnostics-app/src/assets/.gitkeep, What is this for? Can you remove  it, if not going to be used?

The `assets` directory will have all the assets that will be passed on to the build folder as it is. Currently it is empty but will be populated with all the assets (icons, fonts, graphics) that we'll use. Hence we keep the folder with `.gitkeep` which is just an empty file.

> 2. What are the .png and icons? Are these Angular-logo? We need to use icons with appropriate copyrights. Let me know if you need help with this.

They came along with the boilerplate code. Removed them for now. 👍🏻

zerefwayne commented:
> This PR is already large. Can we focus on one or two goals only?

@MahmoudAGawad Most of the files were auto-generated and only a few files were added by me here. I will make sure to make small PRs in future.

MahmoudAGawad commented:
> > This PR is already large. Can we focus on one or two goals only?
> 
> @MahmoudAGawad Most of the files were auto-generated and only a few files were added by me here. I will make sure to make small PRs in future.

But I see changes that can really be moved to next PRs, code logic like:
- Refresh PWA Cache on update
- Add WebComponent support and Lit
- Add Light and Dark theme functionality

If it doesn't break the any part of the app, let's move them, what do you think?

zerefwayne commented:
> > > This PR is already large. Can we focus on one or two goals only?
> > 
> > 
> > @MahmoudAGawad Most of the files were auto-generated and only a few files were added by me here. I will make sure to make small PRs in future.
> 
> But I see changes that can really be moved to next PRs, code logic like:
> 
> * Refresh PWA Cache on update
> * Add WebComponent support and Lit
> * Add Light and Dark theme functionality
> 
> If it doesn't break the any part of the app, let's move them, what do you think?

Okay 👍🏻 I'll try to separate them into three PRs.

zerefwayne commented:
@MahmoudAGawad The goals listed in the PR description were squashed into a single commit before raising this PR. I've separated the other logic into their separate PRs.

zerefwayne committed:
[app] Remove boilerplate code and follow style-guide

- Update copyright comment style /* */ => //
- Update default header
- Remove unused code and files
- Update README.md and Move .editorconfig
- Add jsdoc @fileoverview comments

MahmoudAGawad commented on diagnostics-app/src/app/app.component.css:
```diff
@@ -0,0 +1,10 @@
+/**
+ * Copyright 2021 The Chromium Authors. All rights reserved.
+ * Use of this source code is governed by a BSD-style license that can be
+ * found in the LICENSE file.
+ */
```
use // instead of /* in css files as well.

zerefwayne replied:
/* */ is the only way to add comments in CSS.

https://www.xanthir.com/b4U10 According to this blog, we can have // comments in CSS but it would look like:

```
//Copyright 2021 The Chromium Authors. All rights reserved. {}
//Use of this source code is governed by a BSD-style license that can be {}
//found in the LICENSE file. {}
```

MahmoudAGawad replied:
I looked into chromium files and found out we are using /* as you did. So, let's keep it with a **leading space** preceeding the * at the beginning of each line (starting from the second line)
```
/*
 * Copyright 2021 The Chromium Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */
 ```

MahmoudAGawad commented on diagnostics-app/src/assets/.gitkeep:
```diff
@@ -0,0 +1,3 @@
+// Copyright 2021 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
```
let's add this header for js/ts/css/html files only

MahmoudAGawad commented on diagnostics-app/src/styles.css:
```diff
@@ -0,0 +1,10 @@
+/**
+ * Copyright 2021 The Chromium Authors. All rights reserved.
+ * Use of this source code is governed by a BSD-style license that can be
+ * found in the LICENSE file.
+ */
```
use // instead of /*

zerefwayne replied:
Same as the previous comment.

zerefwayne committed:
[app] Remove copyright header from .gitkeep

zerefwayne committed:
[app] Change css header from /** => /*

MahmoudAGawad reviewed (approved):
Thank you, Aayush. Looks good to me.

Lamzin commented on diagnostics-app/package.json:
```diff
@@ -0,0 +1,42 @@
+    "karma-jasmine-html-reporter": "^1.5.0",
+    "typescript": "~4.2.3"
+  }
+}
```
\n at the end?

zerefwayne replied:
Added. 👍🏻

Lamzin commented on diagnostics-app/angular.json:
```diff
@@ -0,0 +1,112 @@
+            "outputPath": "dist/diagnostics-app",
+            "index": "src/index.html",
+            "main": "src/main.ts",
+            "polyfills": "src/polyfills.ts",
```
Do we need polyfills? Why? :)

zerefwayne replied:
Yes. This file is automatically generated by Angular CLI and was earlier used to support IE and older browsers.

Although we don't need to support them, there are some ES6+ and Typescript features that might not be fully compatible yet. Hence, according to some sources, it is safe to have this file.

Lamzin commented on diagnostics-app/package.json:
```diff
@@ -0,0 +1,42 @@
+    "test": "ng test"
+  },
+  "private": true,
+  "dependencies": {
```
Can you please add a brief comments for each dependency? It will help to understand why do we need them all.

zerefwayne replied:
I've added them, please have a look.

Lamzin commented on diagnostics-app/package.json:
```diff
@@ -0,0 +1,42 @@
+    "@angular/platform-browser-dynamic": "~12.0.2",
+    "@angular/router": "~12.0.2",
+    "@angular/service-worker": "~12.0.2",
+    "rxjs": "~6.6.0",
```
I don't see that it's used somewhere in the code. Do you need it?

zerefwayne replied:
Same as above. Yes, all the dependencies that are listed automatically are used in the code. They're either imported in `app.module.ts` or will be needed very soon.

Lamzin replied:
We should add them when we need them.

zerefwayne replied:
I had a doubt regarding `rxjs` not being used directly. It is being used as a dependency by `@angular/material`

```
"peerDependencies": {
    "@angular/animations": "^12.0.0 || ^13.0.0-0",
    "@angular/cdk": "12.0.3",
    "@angular/core": "^12.0.0 || ^13.0.0-0",
    "@angular/common": "^12.0.0 || ^13.0.0-0",
    "@angular/forms": "^12.0.0 || ^13.0.0-0",
    "rxjs": "^6.5.3"
  }
 ```

So all of them are essential in this PR. 👍🏻

Lamzin replied:
Thanks for double check!

Lamzin commented on diagnostics-app/package.json:
```diff
@@ -0,0 +1,42 @@
+    "@angular/service-worker": "~12.0.2",
+    "rxjs": "~6.6.0",
+    "tslib": "^2.1.0",
+    "zone.js": "~0.11.4"
```
This is used only for testing, should we put it under "dependencies"? Are there some sort of test dependencies?

zerefwayne replied:
- `tslib` is a runtime library for TypeScript that contains all of the TypeScript helper functions.
- `zone.js` is essential to Angular for handling change detections. It allows Angular to know when the UI needs to be refreshed.

So both of them are core essentials of Angular!

Lamzin commented on diagnostics-app/package.json:
```diff
@@ -0,0 +1,42 @@
+  },
+  "private": true,
+  "dependencies": {
+    "@angular/animations": "~12.0.2",
```
Same. Don't see that it's used in the code. Do we need it?

zerefwayne replied:
These were added automatically by the CLI and will be required in the future. For eg. @angular/animations is used in PR #8 for the sidebar module. The are used internally by the modules that are installed.

Lamzin replied:
Sounds good if this required for this PR, but not PR 8

Lamzin commented:
In general looks good, but please double check dependencies, looks like there are some of them we don't need (but I might be wrong about some deps).

zerefwayne committed:
[app] Add newline in package.json

zerefwayne committed:
[app] Add dependency descriptions

Lamzin commented on diagnostics-app/package.json:
```diff
@@ -9,6 +9,22 @@
+    "@angular/platform-broswer": "Supports execution of Angular apps on supported browsers",
+    "@angular/platform-browser-dynamic": "Supports JIT compilation and execution of Angular apps on different supported browsers",
+    "@angular/router": "Implements the Angular Router service , which enables navigation from one view to the next as users perform application tasks",
+    "@angular/service-worker": "This document explains how to enable Angular service worker support in projects that you created with the Angular CLI",
```
`This document` -- why does it refer to some document?

zerefwayne replied:
Fixed it. 👍🏻

Lamzin commented on diagnostics-app/package.json:
```diff
@@ -25,6 +41,20 @@
+    "jasmine-core": "Testing library",
+    "karma": "Test runner",
+    "karma-chrome-launcher": "Used for E2E Testing",
+    "karma-coverage": "Coverage reports for Karma.",
```
nit: please be consistent with dots at the end of comments.

zerefwayne replied:
Thanks for pointing out!

Lamzin reviewed (approved):
LGTM % nits, thanks!

zerefwayne committed:
[app] Fix typos in package.json
